File: sender/Transport/Frame.swift
//
//  Frame.swift
//  SafeHaven — Transport
//
//  Builds the fixed 16-byte little-endian binary media header defined in
//  PROTOCOL §3. Every binary WebSocket frame is this header followed by an
//  opaque payload (H.264 Annex-B access unit or PCM chunk). The header is
//  ALWAYS cleartext; only the payload passes through the encryption seam (§9).
//
//      offset size  field        notes
//      0      1     version      = 0x01
//      1      1     kind         1 = VIDEO_H264, 2 = AUDIO_PCM
//      2      1     flags        bit0 = KEYFRAME (IDR) for video; audio = 0
//      3      1     cacheClass   0 = none, 1 = retain-last-of-kind
//      4      4     seq          uint32 LE, per-kind monotonic from the sender
//      8      8     ptsMicros    uint64 LE, presentation time in microseconds
//      16     ...   payload      encoded access unit OR PCM chunk
//

import Foundation

/// Media stream kind (header byte[1]). PROTOCOL §3.
enum FrameKind: UInt8 {
    case video = 1   // VIDEO_H264
    case audio = 2   // AUDIO_PCM
}

/// Header flag bits (header byte[2]). PROTOCOL §3.
struct FrameFlags: OptionSet {
    let rawValue: UInt8
    /// bit0 — set on H.264 IDR (keyframe) access units; never set for audio.
    static let keyframe = FrameFlags(rawValue: 1 << 0)
}

/// Cache class (header byte[3]) — instructs the relay's media cache. PROTOCOL §3/§7.
enum CacheClass: UInt8 {
    case none = 0            // do not retain (audio, delta video)
    case retainLastOfKind = 1 // retain-last-of-kind (latest IDR only)
}

/// The protocol version byte (header byte[0]).
private let kFrameVersion: UInt8 = 0x01

/// Length of the fixed binary header in bytes.
let kFrameHeaderLength = 16

enum Frame {

    /// Build a complete binary frame: 16-byte LE header + payload.
    ///
    /// - Parameters:
    ///   - kind: VIDEO_H264 or AUDIO_PCM.
    ///   - flags: keyframe flag for IDR video; empty otherwise.
    ///   - cacheClass: retain-last-of-kind on IDR; none otherwise.
    ///   - seq: per-kind monotonic sequence (resets on sender reconnect).
    ///   - ptsMicros: capture-clock presentation timestamp in microseconds.
    ///   - payload: the opaque (already-transformed) payload bytes.
    static func make(kind: FrameKind,
                     flags: FrameFlags,
                     cacheClass: CacheClass,
                     seq: UInt32,
                     ptsMicros: UInt64,
                     payload: Data) -> Data {
        var data = Data(capacity: kFrameHeaderLength + payload.count)

        // byte[0] version
        data.append(kFrameVersion)
        // byte[1] kind
        data.append(kind.rawValue)
        // byte[2] flags
        data.append(flags.rawValue)
        // byte[3] cacheClass
        data.append(cacheClass.rawValue)
        // bytes[4..8) seq, little-endian u32
        appendLittleEndian(&data, seq)
        // bytes[8..16) ptsMicros, little-endian u64
        appendLittleEndian(&data, ptsMicros)
        // bytes[16..] payload
        data.append(payload)

        return data
    }

    // MARK: - Little-endian writers

    private static func appendLittleEndian(_ data: inout Data, _ value: UInt32) {
        var le = value.littleEndian
        withUnsafeBytes(of: &le) { data.append(contentsOf: $0) }
    }

    private static func appendLittleEndian(_ data: inout Data, _ value: UInt64) {
        var le = value.littleEndian
        withUnsafeBytes(of: &le) { data.append(contentsOf: $0) }
    }
}

File: receiver/src/transport/frame.ts
/**
 * frame.ts — binary media frame parsing (PROTOCOL §3).
 *
 * Every WS BINARY frame is a fixed 16-byte LITTLE-ENDIAN header followed by an
 * opaque payload. The header is ALWAYS cleartext; the payload is the part the
 * encryption seam (§9) will transform later (see inboundTransform.ts).
 *
 *   offset size  field        notes
 *   0      1     version      = 0x01
 *   1      1     kind         1 = VIDEO_H264, 2 = AUDIO_PCM
 *   2      1     flags        bit0 = KEYFRAME (IDR) for video; 0 for audio
 *   3      1     cacheClass   0 = none, 1 = retain-last-of-kind
 *   4      4     seq          uint32, per-kind monotonic (resets on sender reconnect)
 *   8      8     ptsMicros    uint64, presentation timestamp in microseconds
 *   16     ...   payload      encoded video access unit OR PCM chunk (opaque)
 */

export const FRAME_VERSION = 0x01;
export const HEADER_BYTES = 16;

/** Media kind discriminator (header byte[1]). */
export enum MediaKind {
  Video = 1, // VIDEO_H264
  Audio = 2, // AUDIO_PCM
}

/** flags bit0 — set on video IDR access units. */
export const FLAG_KEYFRAME = 0x01;

/** A parsed binary media frame. `payload` is a view INTO the original buffer. */
export interface MediaFrame {
  version: number;
  kind: MediaKind;
  /** True when flags.bit0 is set (video IDR). Always false for audio. */
  keyframe: boolean;
  /** 0 = none, 1 = retain-last-of-kind. */
  cacheClass: number;
  /** Per-kind monotonic sequence; resets on sender reconnect. */
  seq: number;
  /** Presentation timestamp in microseconds (capture clock). */
  ptsMicros: number;
  /**
   * Opaque payload bytes. This is a Uint8Array VIEW over the source
   * ArrayBuffer (no copy). Consumers that retain it past the current tick
   * should copy if the source buffer may be reused.
   */
  payload: Uint8Array;
}

/**
 * Parse a 16-byte LE header + payload out of an ArrayBuffer received on the
 * WebSocket. Returns null if the buffer is too short or the version byte is
 * unrecognized (defensive — a malformed frame must never crash the receiver).
 *
 * ptsMicros is read with DataView.getBigUint64 then narrowed to a JS number.
 * Microsecond timestamps stay well within Number.MAX_SAFE_INTEGER for any
 * realistic capture clock, so the narrowing is lossless in practice.
 */
export function parseMediaFrame(buffer: ArrayBuffer): MediaFrame | null {
  if (buffer.byteLength < HEADER_BYTES) return null;

  const view = new DataView(buffer);
  const version = view.getUint8(0);
  if (version !== FRAME_VERSION) return null;

  const kind = view.getUint8(1) as MediaKind;
  const flags = view.getUint8(2);
  const cacheClass = view.getUint8(3);
  // little-endian for ALL multi-byte fields (§3).
  const seq = view.getUint32(4, /* littleEndian */ true);
  const ptsMicros = Number(view.getBigUint64(8, /* littleEndian */ true));

  // Payload is everything after the header. Zero-copy view.
  const payload = new Uint8Array(buffer, HEADER_BYTES, buffer.byteLength - HEADER_BYTES);

  return {
    version,
    kind,
    keyframe: (flags & FLAG_KEYFRAME) !== 0,
    cacheClass,
    seq,
    ptsMicros,
    payload,
  };
}
